pkg/resource/deploy/step.go

```diff
@@ -337,26 +337,34 @@ func (s *DeleteStep) Logical() bool           { return !s.replacing }
 func (s *DeleteStep) Apply(preview bool) (resource.Status, StepCompleteFunc, error) {
 	// Refuse to delete protected resources (unless we're replacing them in
 	// which case we will of checked protect elsewhere)
-	if !s.replacing && s.old.Protect {
+	if !s.replacing && (s.old.Protect || s.old.DeleteBehaviour == resource.DeleteBehaviourProtect) {
 		return resource.StatusOK, nil, fmt.Errorf("unable to delete resource %q\n"+
 			"as it is currently marked for protection. To unprotect the resource, "+
 			"either remove the `protect` flag from the resource in your Pulumi "+
 			"program and run `pulumi up` or use the command:\n"+
 			"`pulumi state unprotect '%s'`", s.old.URN, s.old.URN)
 	}
 
-	// Deleting an External resource is a no-op, since Pulumi does not own the lifecycle.
-	if !preview && !s.old.External {
-		if s.old.Custom {
-			// Invoke the Delete RPC function for this provider:
-			prov, err := getProvider(s)
-			if err != nil {
-				return resource.StatusOK, nil, err
-			}
+	if preview {
+		// Do nothing in preview
+	} else if s.old.External {
+		// Deleting an External resource is a no-op, since Pulumi does not own the lifecycle.
+	} else if s.old.DeleteBehaviour == resource.DeleteBehaviourDrop {
+		// Deleting a "drop on delete" is a no-op as the user has explicitly asked us to not delete the resource.
+		// But we want to diag warn that Pulumi has stopped tracking this resource
+		msg := fmt.Sprintf("Reference to %s dropped, Pulumi has not deleted this resource", s.old.ID)
```

Comment:
Open to better wording on this message. Should it be even more clear about the cost implications?

Reply:
Yup this may be a bit tricky. "Pulumi dropped the reference to %s is no longer tracking it, but it remains provisioned in the cloud"? Except it's not necessarily cloud. "remains provisioned?"